onnxruntime/core/optimizer/qdq_transformer/qdq_final_cleanup.cc

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/qdq_transformer/qdq_final_cleanup.h"
+
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/qdq_transformer/qdq_util.h"
+#include "core/optimizer/selectors_actions/actions.h"
+#include "core/optimizer/utils.h"
+
+namespace onnxruntime {
+Status QDQFinalCleanupTransformer::ApplyImpl(Graph& graph, bool& modified, int graph_level,
+                                             const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+  for (auto node_index : node_topology_list) {
+    auto* node_ptr = graph.GetNode(node_index);
+    if (node_ptr == nullptr)
+      continue;  // node was removed as part of an earlier optimization
+
+    Node& q_node = *node_ptr;
+    ORT_RETURN_IF_ERROR(Recurse(q_node, modified, graph_level, logger));
+
+    if (!QDQ::MatchQNode(q_node) ||
+        !graph_utils::IsSupportedProvider(q_node, GetCompatibleExecutionProviders()) ||
+        !optimizer_utils::CheckOutputEdges(graph, q_node, 1)) {
```

Comment:
>!optimizer_utils::CheckOutputEdges(graph, q_node, 1)) { [](http://example.com/codeflow?start=8&length=55)

do we need to cover the corner case that the output of Q is graph output?

Reply:
CheckOutputEdges does that I believe. One edge, no graph output for that to return true.